query/src/main/java/org/infinispan/query/impl/PartitionHandlingSupport.java

```diff
@@ -0,0 +1,39 @@
+package org.infinispan.query.impl;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.configuration.cache.ClusteringConfiguration;
+import org.infinispan.partitionhandling.AvailabilityMode;
+import org.infinispan.partitionhandling.PartitionHandling;
+import org.infinispan.query.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * @since 9.3
+ */
+public final class PartitionHandlingSupport {
+
+   private static final Log LOGGER = LogFactory.getLog(CacheQueryImpl.class, Log.class);
+
+   private final boolean isClustered;
+   private final PartitionHandling partitionHandling;
+   private final AdvancedCache<?, ?> cache;
+
+   public PartitionHandlingSupport(AdvancedCache<?, ?> cache) {
+      this.cache = cache;
+      ClusteringConfiguration clusteringConfiguration = cache.getCacheConfiguration().clustering();
+      this.isClustered = clusteringConfiguration.cacheMode().isClustered();
+      this.partitionHandling = isClustered ? clusteringConfiguration.partitionHandling().whenSplit() : null;
+   }
+
+   public void checkCacheAvailable() {
+      if (!isClustered) return;
+
+      AvailabilityMode availability = cache.getAvailability();
+      if (availability == AvailabilityMode.AVAILABLE) return;
+
+      if (partitionHandling != PartitionHandling.ALLOW_READ_WRITES) {
```

Comment:
Pasting from the chat the reason why I changed:
"I see, I confess I am in doubt about this one since this config revolves around keys ownership, and query is value oriented. Also, query data can be located on disk and has no info about segments whatsoever...
So ALLOW_READ could allow a query that results entries whose keys in not in the partition, thus contradicting the javadoc for ALLOW_READS"

Reply:
> Also, query data can be located on disk and has no info about segments whatsoever...
So ALLOW_READ could allow a query that results entries whose keys in not in the partition, thus contradicting the javadoc for ALLOW_READS"

Is it possible that all the data needed to answer the query is in the index, and the entities don't have to be fetched with `getAll()`? Otherwise `getAll()` would also throw an `AvailabilityException` in this case.

It's probably still safer to throw the exception for `ALLOW_READS`, I'd just like to see it documented better.

Reply:
And my reply to that was roughly something like: let''s do exactly what we do for expiration + query. Keys returned from index that have a null value in current partition during fetch will be just quietly discarded from query results.  So the returned results will only contain data from current partiion, which obeys the  ALLOW_READS expectations.

Reply:
>  Is it possible that all the data needed to answer the query is in the index, and the entities don't have to be fetched with getAll()

Only if the query is empty, otherwise it always need to go through getAll

Reply:
> Keys returned from index that have a null value

Is that what happens with ALLOW_READ for a key that is not in the partition? It returns null? Or does it thrown an Exception? Which one?

Reply:
My feeling is 99% it will throw AvailabilityException

Reply:
So filtering out nulls is out of question then. But , what if the query does not not hit any faulty key? Then no exception, and the query results are correct.

Reply:
>  the query results are correct

Not sure, as writes are not allowed in a ALLOW_READ partition

Reply:
If no writes are allowed, then no index is updated. The data and the index are just as stale. 

My whole point of view regarding allowing a query to proceed in ALLOW_READ case (and possibly fail later) instead of failing early is to have the same behaviour as If I get the data by key, given I know the keys I want. A query that returns the same keys should have the same behaviour.  The fail fast approach is only mandatory for DENY_R_W case, imho

Reply:
I needed a refresh on this as well: I thought with `ALLOW_READS` degraded mode reads throw an `AvailabilityException` if the local partition doesn't contain ALL the owners of the key, but it turns out it only throws the `AvailabilityException` if the local partition doesn't contain ANY of the key owners.

@ryanemerson if there's any check to prevent reads in a minority partition when the majority partition stays available and allows writes?

Anyway, even if `ALLOW_READS` allows stale reads in the minority partition, the important point is that it always returns some value (possibly stale) or throws an `AvailabilityException`, it never returns `null` for unavailable keys.

Reply:
I'm fine with that. If it returns a stale value, it is just as stale as the index. It is actually in sync with the index, but both data and index are behind majority partition. If it throws availabilityexception, also fine. No nulls, even better.

So with ALLOW_READS we'll either have a search result that is in sync with the index, or an availabilityexception. It's a best effort attempt. IMHO it is better than failing right from the start.

"from Person p where p.name='Dan'" should yield the same result (or exception) as personCache.get("Dan");

Reply:
@anistor has a point. 
@danberindei to reach a consensus on this, could you elaborate why you "tend to agree that we shouldn't allow queries in degraded mode when configured with ALLOW_READS"?

Reply:
@gustavonalle I had 2 things in mind:

# I thought that `ALLOW_READS` reads throws `AvailabilityException` any time one of the owners is missing, so most would eventually throw an AE during iteration.
# I thought queries using projections might skip the `getAll()` call and return data from the index, meaning some queries would throw the AE during iteration and some wouldn't. And I wasn't thinking about queries that return 0 results :)

I no longer believe either of those, so I now agree with @anistor

Reply:
ok, so we need to make sure that no query is resolved solely through the index without touching the cache. From the top of my head:

1) if any field of an object is marked as ```Store.YES``` and thus the value persisted in the index, and a query has projection to only this field, it's not necessary to touch the cache data. 
Whether this happens in our query engine, I'd need to verify.

2) Count queries, when using ```getResultSize()``` on an indexed query, does not need to touch the cache.

So now I disagree that we should do queries on ALLOW_READS  :)

Reply:
Which makes me think we should not support indexing or do not use the index with PH. Only non-indexed query should be allowed, because the index is degraded :)

Reply:
So what are the conditions to get an AvailabilityException with ALLOW_READS ?

Reply:
Should we just assume that query does not support PH and throw AvailabilityException for all cases?  We can allow queries for degraded partitions on a per-case basis in another JIRA...

Reply:
> So what are the conditions to get an AvailabilityException with ALLOW_READS ?

Try to perform a write on a partition that is degraded.

Reply:
@anistor reads will also throw an AvailabilityException if none of the owners are in the local partition.

Reply:
@anistor Sorry I should have elaborated. Dan is correct, for completeness an AvailabilityException will also occur if a partition does not contain all owners of an entry and the `FORCE_WRITE_LOCK` flag is set on a read operation as it's not possible to acquire all of the locks.